FREYA-1430: integrate direct WordPress API for job listings

misterabdullahAziz opened the pull request:
### 📋 Summary
<!-- Briefly describe WHAT and WHY of this PR -->

This pull request refactors the job listing functionality on the `layouts/jobs/list.html` page. It introduces a new data-fetching mechanism that integrates directly with the SciLifeLab WordPress API, improves filtering and sorting performance, and simplifies the user interface by removing modal-based job descriptions. Key changes include API integration, optimised filtering logic, and UI updates.

### 🛠️ Changes Made
<!-- One-liners or bullet points -->
- API Integration and Data Transformation:
-- Replaced the previous XMLHttpRequest-based method with a `fetch` call to retrieve job data directly from the SciLifeLab WordPress API. Added logic to filter, transform, and format API data for display.
-- Enhanced job type and employer extraction with fallback logic and optimised detection based on job titles and categories.
- Filtering and Sorting Improvements:
-- Improved filtering logic by caching DOM elements and using more efficient iteration methods for job type and employer filters.
-- Replaced arrays with `Set` objects for job types and employers to ensure uniqueness and improve performance.
- UI Updates:
-- Removed modal-based job descriptions and replaced them with direct links to job postings, simplifying the user interface.
-- Updated the job type and employer filters to use sorted lists derived from `Set` objects for better organisation.
-- Adjusted the "See Details" button to directly link to the job posting instead of opening a modal.
-  Code Simplification:
-- Consolidated the job fetching and filtering initialisation into a single callback function to streamline the process.
 
### 🔍 Notes for Reviewers
<!-- Tips, edge cases, or test instructions -->
The modal popup was previously used to display a "brief description" field that was manually added by human intervention. However, this approach has several limitations:

1. **Manual Maintenance Burden**: The brief descriptions required ongoing human effort to create and maintain
2. Description Quality Issues: Initially tried a simple approach showing 250 characters from job content, but this resulted in wage information and poor user experience.
3. **API Integration Challenges**: Attempted to integrate with [deepseek.scilifelab](https://deepseek.scilifelab.se/) APIs for automatic description generation, but this resulted in:
   - **30+ minute processing times** for job descriptions
   - **Unacceptable user experience** with cards taking too long to display
   - **Practical infeasibility** for real-time job listings

#### Solution: Direct Job Links
After team discussion in standup, we decided to:
- **Remove the brief description field entirely** from job cards
- **Eliminate modal popups** that were only showing this limited information
- **Provide direct links** to the actual job postings instead

The job cards now provide essential information (title, employer, deadline, job type) with a direct link to the complete job posting, offering users the most up-to-date and comprehensive information available.

### ✅ Checklist
- [x] PR title follows the pattern: `FREYA-XXXX: Clear and short description`
- [x] Jira / Github issue is linked
- [x] Assignee is selected
- [x] Code and content adhere to [conventions](https://scilifelab.atlassian.net/wiki/spaces/CDP2/pages/1909424133/Guidelines+for+the+portal+content)
- [x] Automated checks pass
- [x] Reviewer is selected when the PR is marked as ready for review

### 🔗 Jira Issue
<!-- Example: FREYA-914 -->
Closes: [FREYA-1430](https://scilifelab.atlassian.net/browse/FREYA-1430)

misterabdullahAziz committed:
feat(jobs): integrate direct WordPress API for job listings

senthil10 commented on layouts/jobs/list.html:
```diff
@@ -112,35 +112,123 @@ <h5 class="modal-title" id="suggestionModalLabel">Jobs, PhD positions, Postdoc p
-        loadend_callback();
-      };
-      req.send(null);
+    function getDataFromUrl(url, load_callback) {
```
This can be a function that just reads URL and return the data, no need for load_callback with `fetch`

senthil10 commented on layouts/jobs/list.html:
```diff
@@ -112,35 +112,123 @@ <h5 class="modal-title" id="suggestionModalLabel">Jobs, PhD positions, Postdoc p
-      req.send(null);
+    function getDataFromUrl(url, load_callback) {
+      // Fetch jobs directly from SciLifeLab WordPress API
+      fetch('https://www.scilifelab.se/wp-json/wp/v2/career?orderby=archive_date&per_page=50')
```
This can be passed as `url` while calling the function

senthil10 commented on layouts/jobs/list.html:
```diff
@@ -112,35 +112,123 @@ <h5 class="modal-title" id="suggestionModalLabel">Jobs, PhD positions, Postdoc p
+            }
+
+            // Extract employer
+            let employer = '';
```
If default is `SciLifeLab`, then we can set it here and overwrite it below if conditions satisfy ?

senthil10 commented on layouts/jobs/list.html:
```diff
@@ -112,35 +112,123 @@ <h5 class="modal-title" id="suggestionModalLabel">Jobs, PhD positions, Postdoc p
+            }
+
+            // Extract job type from API or use default
+            let jobType = [];
```
With the automation way we don't need it to be a array, it can be a string and start it with the default value and it can be overwritten below.

senthil10 commented on layouts/jobs/list.html:
```diff
@@ -112,35 +112,123 @@ <h5 class="modal-title" id="suggestionModalLabel">Jobs, PhD positions, Postdoc p
+            };
+          });
+
+          const result = { items: transformedJobs };
```
Again the `items` where just effects of copy paste, this function return the array

senthil10 commented on layouts/jobs/list.html:
```diff
@@ -112,35 +112,123 @@ <h5 class="modal-title" id="suggestionModalLabel">Jobs, PhD positions, Postdoc p
+        })
+        .catch(error => {
+          console.error('Error fetching jobs:', error);
+          var req = new XMLHttpRequest();
```
The following is not necessary as the idea is to replace the current approach 👍

senthil10 commented on layouts/jobs/list.html:
```diff
@@ -310,37 +350,54 @@ <h5 class="modal-title" id="descriptionModalLabel${ind}">${job.title}</h5>
+            
+            var showJob = false;
+            
+            if (!job_type.length && !employer.length) {
```
This is already checked in line 363, so this will always evaluate false here thus not needed here.

senthil10 commented on layouts/jobs/list.html:
```diff
@@ -350,8 +407,10 @@ <h5 class="modal-title" id="descriptionModalLabel${ind}">${job.title}</h5>
+      function(jobsData) {
+        addJobs(jobsData);
+        enableFiltering();
+      }
```
As mentioned, the idea is to stop using manually produced data in blob server and use direct approach. Also the new calling could be something like 

```
async function getDataFromUrl(url) { ....... }

async function addJobs (dataArray) { ....... }

function enableFiltering() { ........ }

document.addEventListener("DOMContentLoaded", async () => {
    const availableJobs = await getDataFromUrl('https://www.scilifelab.se/wp-json/wp/v2/career?orderby=archive_date&per_page=50');
    await addJobs(availableJobs);
    enableFiltering();
})
```

senthil10 reviewed (changes requested):
Great work, had some comments.

Also, looks like it takes 2-3 seconds to load the jobs, so it is good to have a spinner says fetching jobs until it is loaded

misterabdullahAziz committed:
refactor(jobs): streamline job data fetching and processing

misterabdullahAziz committed:
feat(jobs): add loading indicator for job listings

senthil10 commented on layouts/jobs/list.html:
```diff
@@ -310,49 +360,64 @@ <h5 class="modal-title" id="descriptionModalLabel${ind}">${job.title}</h5>
+              showJob = true;
+            } else if (job_type.length && employer.length) {
+              // Both filters active
+              showJob = job_type.some(type => jobType.toLowerCase() === type.toLowerCase()) && 
```
`jobType` will not be a list anymore, so maybe can directly use `.includes` like for `employer` ?

misterabdullahAziz replied:
yes, I remember now. `jobType` is string true. But `job_type` is not string instead it is array based on checkbox checked in the filter. moreover it is also checking for case-sensitivity within check which cannot be done within includes.
Anyhow I can make it consistence.

senthil10 reviewed (commented):
Looks good, just 2 very minor nitpicking comments 😄 

We can discuss when Liane is back and then can be merged

misterabdullahAziz committed:
refactor(jobs): enhance job filtering logic for improved performance and clarity

senthil10 commented on layouts/jobs/list.html:
```diff
@@ -104,43 +104,137 @@ <h5 class="modal-title" id="suggestionModalLabel">Jobs, PhD positions, Postdoc p
+            jobType = 'Researcher or Engineer';
+          } else if (category.includes('technical-and-administrative-personnel')) {
+            jobType = 'Admin and Technical';
+          }
```
@misterabdullahAziz Forgot about this to discuss it with @LianeHughes

senthil10 reviewed (approved):
@misterabdullahAziz Great work, code looks nice and better.

**NOTE:** Before merging, discuss/confirm the job type part wlth @LianeHughes once.

misterabdullahAziz commented:
Oh, missed opportunity to discuss it today. We can discuss it later maybe.

misterabdullahAziz commented:
Oh, missed opportunity to discuss it today. We can discuss it later maybe.

senthil10 reviewed (changes requested):
While fixing the Job types as discussed, also remove the suggestion form and related JS

senthil10 commented on layouts/jobs/list.html:
```diff
@@ -104,43 +104,137 @@ <h5 class="modal-title" id="suggestionModalLabel">Jobs, PhD positions, Postdoc p
-      };
-      req.send(null);
+    // Constants
+    const JOBS_API_URL = 'https://www.scilifelab.se/wp-json/wp/v2/career?orderby=archive_date&per_page=50';
```
We will only show `Data Centre` jobs, so the `URL` should be 

```
https://www.scilifelab.se/wp-json/wp/v2/career?categories=980&orderby=archive_date&per_page=50
```

misterabdullahAziz committed:
refactor(jobs): remove job suggestion modal and update API URL for job listings

misterabdullahAziz commented:
I fixed jobs url and removed suggestion form.
Unfortunately I could not get any response with this API now that has not passed archive date.
About type, i tried to figure out what we have decided but could not remember :(

senthil10 commented:
Yes, currently no open jobs marked/tagged with DC, so it is expected to not have any listed.

For type, we said initially it can just be something like `jobType = job.acf?.job_type || "Job"` and later if needed, we can update it.

misterabdullahAziz commented:
but no job has any property `job_type`
are we planning to add it in wordpress APIs?

senthil10 commented:
> but no job has any property `job_type` are we planning to add it in wordpress APIs?

I just re-iterated whatever you mentioned in `determineJobType` function 😅 (any reason you used a non-exisiting property ?). 

Anyway in that case we have to try to guess it from the job title like you do, but only check for the designations that already exist in [DC](https://www.scilifelab.se/contact/data-centre/)

misterabdullahAziz committed:
refactor(jobs): remove job type determination logic and related UI elements

misterabdullahAziz committed:
refactor(jobs): update job listings section to focus on DC jobs

senthil10 commented on layouts/jobs/list.html:
```diff
@@ -5,149 +5,110 @@
-                          life science. If you know of any relevant jobs or project opportunities that should be shown 
-                          here, please suggest them to us using the "Tell us about an open position" button above.
+                          We are currently unaware of any upcoming job opportunities at the SciLifeLab Data Centre. 
+                          All positions shown here are specifically from the Data Centre team.
```
I think the text can just say _"Currently there are no job opportunities at the SciLifeLab Data Centre"_

This site will represent the information from DC, so won't be good to say we are unaware of ourself 😄

misterabdullahAziz committed:
refactor(jobs): update no jobs message for clarity and styling

senthil10 reviewed (approved):
Looks good to me, if @LianeHughes is happy with the texts can be merged

LianeHughes commented on content/jobs/_index.md:
```diff
@@ -16,6 +16,4 @@ aliases:
-In this section, we show the latest job opportunities, PhD positions, postdoc positions and masters project opportunities related to *data-driven life science in Sweden*. We may show some 'global' opportunities but, in each case, it will be possible for individuals based in Sweden to apply.
-
-If you know of any opportunity that is not listed below, but feel that it should be, please *click the button below and fill in the short form*.
+In this section, we show the latest job opportunities from the **SciLifeLab Data Centre**. These positions focus on IT infrastructure, data management, and computational services that support life science research across Sweden.
```
```suggestion
Here are the latest job opportunities within SciLifeLab Data Centre. Positions typically relate to open science, research data management, and developing and managing IT infrastructure. 
```

LianeHughes commented on layouts/jobs/list.html:
```diff
@@ -5,149 +5,109 @@
-                          life science. If you know of any relevant jobs or project opportunities that should be shown 
-                          here, please suggest them to us using the "Tell us about an open position" button above.
+        var jobsHTML = `<div class="alert alert-secondary mt-4 text-center">
+                          Currently there are no job opportunities at the SciLifeLab Data Centre.
```
```suggestion
                          There are currently no open job opportunities at SciLifeLab Data Centre.
```

LianeHughes reviewed (commented):
Minor text updates. 
Please let me know what information needs to be added on the web side for jobs to be visible here, I can then inform the relevant parties.

misterabdullahAziz committed:
refactor(jobs): update job description and no jobs message for clarity

LianeHughes reviewed (approved):
Looks good to me (on the text element, I see that all else is pre-approved, so I did not see a need to reassess)
